tardis/plasma/properties/nlte_rate_equation_solver.py

```diff
@@ -0,0 +1,290 @@
+    outputs = ("ion_number_density_nlte", "electron_densities_nlte")
+
+    def calculate(
+        self,
+        gamma,
+        alpha_sp,
+        alpha_stim,
+        coll_ion_coeff,
+        coll_recomb_coeff,
+        partition_function,
+        levels,
+        level_boltzmann_factor,
+        phi,
+        rate_matrix_index,
+        number_density,
+    ):
+
+        (
+            photo_ion_rate,
+            rad_recomb_rate_coeff,
+            coll_ion_coefficient,
+            coll_recomb_coefficient,
+        ) = NLTERateEquationSolver.prepare_ion_recomb_rates_nlte_ion(
+            gamma,
+            alpha_sp,
+            alpha_stim,
+            coll_ion_coeff,
+            coll_recomb_coeff,
+            partition_function,
+            levels,
+            level_boltzmann_factor,
+        )
+
+        # >>>TODO:initial electron density should be included in the initial guess, added in a future PR
+        initial_electron_density = number_density.sum(axis=0)
+        # <<<
+        rate_matrix = NLTERateEquationSolver.calculate_rate_matrix(
+            phi[0],
+            initial_electron_density[0],
+            rate_matrix_index,
+            photo_ion_rate[0],
+            rad_recomb_rate_coeff[0],
+            coll_ion_coefficient[0],
+            coll_recomb_coefficient[0],
+        )
+        return -1, -1
+
+    @staticmethod
+    def calculate_rate_matrix(
+        phi_shell,
+        electron_density,
+        rate_matrix_index,
+        photo_ion_rate,
+        rad_recomb_rate_coeff,
+        coll_ion_coefficient,
+        coll_recomb_coefficient,
+    ):
+        """
+
+        Parameters
+        ----------
+        phi_shell : DataFrame
+            Saha Factors in the current shell
+        electron_density : float
+            Guess for electron density in the current shell
+        rate_matrix_index : MultiIndex
+            Index used for constructing the rate matrix
+        photo_ion_rate : DataFrame
+            Photo ionization rates
+        rad_recomb_rate_coeff : DataFrame
+            Radiative recombination coefficients(should get multiplied by electron density)
+        coll_ion_coefficient : DataFrame
+            Collisionional ionization coefficients(should get multiplied by electron density)
+        coll_recomb_coefficient : DataFrame
+            Collisionional recombination coefficients(should get multiplied by electron density^2)
+
+        Returns
+        -------
+        DataFrame
+            Rate matrix used for nlte solver.
+        """
+        rate_matrix = pd.DataFrame(
+            0.0, columns=rate_matrix_index, index=rate_matrix_index
+        )
+        rad_recomb_rates = rad_recomb_rate_coeff * electron_density
+        coll_ion_rates = coll_ion_coefficient * electron_density
+        coll_recomb_rates = coll_recomb_coefficient * electron_density**2
+        atomic_numbers = (
+            rate_matrix_index.get_level_values(0).unique()[:-1].values
```

Comment:
Instead of cutting of the last element, I would do `rate_matrix_index.drop('ne')`. In my opinion, this is safer (and also easier for others to follow).